[DataPipe] Revamp serialization logic of DataPipes

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74984

This PR removes the custom serialization logic (i.e. try with `pickle` then with `dill`) inside `__getstate__` and `__setstate__` of DataPipe classes because it causes circular reference and exceptions. Instead, the attempt to use `dill` instead of `pickle` is moved to the `traversal` function for DataPipe graphs and will be included within DataLoader2 as well.

As a result, some tests need to be changed (to use `dill` explicitly when necessary) and `serialization.py` has mostly been cleaned up as the functions are no longer needed.

This should fix https://github.com/pytorch/data/issues/237. 

Differential Revision: [D36111757](https://our.internmc.facebook.com/intern/diff/D36111757)

NivekT committed:
[DataPipe] Restrict special serialization for non-method functions

[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -93,11 +94,10 @@ def class_function(cls, enable_df_api_tracing, source_dp, *args, **kwargs):
             state_dict = {}
             for k, v in self.__dict__.items():
-                if callable(v):
+                if isfunction(v) and not ismethod(v):
```
By not applying the special serialization to methods, the example given @pmeier is not an issue. In general, I think it is fair to assume that methods written in a `IterDataPipe` should be serializable by `pickle`, otherwise a custom `__getstate__` should be provided in the class as well.

I don't think any method directly uses `lambda` so I believe this should work. Any edge cases that you can think of?

ejguan replied:
How about the method not bounded to `IterDataPipe`?
If the method is bounded into a function, I am not sure if this is possible to pickle. If not, we have to try using dill.

pmeier replied:
I think the issue only happens for methods bound to `IterDataPipe`'s, right? Maybe we can be strict here to avoid other issues.

NivekT replied:
I believe that with this condition `isfunction(v) and not ismethod(v)` here, only non-method function within a `IterDataPipe` will go through the custom serialization function `serialize_fn`.

This should not impact any method or other `IterDataPipe`. Do you have an example where this may not work?

NivekT replied:
@ejguan @pmeier @VitalyFedyunin Checking if people have any thoughts on this

ejguan replied:
Let's list all use cases of this function:
1. traverse with `only_datapipe=True`
2. traverse with `only_datapipe=False`
3. multiprocessing

For scenario 2 and 3, we need to pickle/dill functions or methods. Then, I have a noob question. I believe both `pickle` and `dill` are able to handle circular reference by themselves, could we simply using `dill` when it's available during `traverse` and `multiprocessing`.

Not sure if it's possible for `multiprocessing` though.

NivekT replied:
Essentially, you are proposing a backend switch (i.e. always use `dill` when available, and don't return a tuple of serialized object and serialization type)?

I can see that working for scenario 2, but it seems like `multiprocessing` will still use `pickle` unless `pickle.Pickler` is monkey-patched (seems too hacky). But I'm also a noob on this, anyone else have a better suggestion?

I would think by editing the `__getstate__` and `__setstate__` of the class as we do here, `dill` should be properly used in the right situation when available.

NivekT replied:
I just updated the logic to check if `fn` is a lambda rather than trying to `pickle.dumps` `fn` first.

I also added this test case that has a lambda attribute (and uses it) and the test passes without any recursive error as long as `dill` is installed`.
```python
class LambdaIterDataPipe(IterDataPipe):
        def add_one(self, x):
            return x + 1

        def classify(self, x):
            return 0

        def __init__(self, fn, source_dp=None):
            self.fn = fn
            self.lambda_fn = lambda x: x + 1
            self.source_dp = source_dp if source_dp else dp.iter.IterableWrapper([1, 2, 4])
            self._dp = self.source_dp.map(self.add_one).map(self.lambda_fn).demux(2, self.classify)[0]

        def __iter__(self):
            yield from self._dp
```
```python
self.assertTrue(list(TestCircularSerialization.LambdaIterDataPipe(lambda x: x + 1)) ==
                list(dill.loads(dill.dumps(TestCircularSerialization.LambdaIterDataPipe(lambda x: x + 1)))))
_ = traverse(TestCircularSerialization.LambdaIterDataPipe(fn=_fake_fn), only_datapipe=True)
_ = traverse(TestCircularSerialization.LambdaIterDataPipe(fn=_fake_fn), only_datapipe=False)
```

ejguan replied:
Just want to double check if it works when the function is embedded into a nested data structure?
```
self.lambda_fn = [lambda x: x + 1, ]
```
This lambda function itself won't be handled by the special serialization. Am I right?

NivekT replied:
Yes, that is correct. But if `dill.dumps` is used, it will still work fine.

I think it is worth re-stating the scope of the problem that we are attempting to solve here.

At first, we were only serializing certain function arguments of built-in DataPipes (e.g. `fn` that get passed into `Mapper`), as you can see the code on the left-hand side in [this PR](https://github.com/pytorch/pytorch/pull/72896/files). When `dill` is available, we serialize/deserialize with `dill` for those `fn` within built-in DataPipe.

In this PR, we are also targeting something similar - i.e. non-method functions within DataPipes. I think that is a reasonable coverage since it covers all the possible lambdas that can exist within the built-in DataPipes. (I'm going to clean up this PR slightly and add a comment to make this intention more explicitly).

If the user chooses to do something more unusual, I think they should either 1) write their own `__getstate__` and `__setstate__` in their custom `IterDataPipe` (as we do for `demux` and `fork` to avoid storing the buffer) or 2) use `dill.dumps`.

NivekT replied:
That said, I think it will be useful to have a backend switch within DataLoader2 (specifically allowing users to choose to serialize with `pickle` or `dill`). Thoughts?

ejguan replied:
I think the reason we bring the serialization logic from each DataPipe instance to this base class is for simplicity and generalization. There are too many corner cases we didn't consider at the beginning. If we want to let users to do custom `__getstate__` and `__setstate__`, why not we move the special serialization logic to the required `DataPipe` like `Mapper`, `Filter`, etc. And, let them to handle `lambda` using `dill` if available.

> That said, I think it will be useful to have a backend switch within DataLoader2 (specifically allowing users to choose to serialize with `pickle` or `dill`). Thoughts?

This is reasonable.

NivekT replied:
What other common corner cases do you think we should handle here assuming we will add a backend switch within DataLoader2? A `list` of unpicklable objects?

ejguan replied:
For example of the `DataPipe` with `buffer`, we would never know the element type within the `buffer` when we want to do serialization/checkpoint in the middle of epoch.

NivekT replied:
Right, and current that is handled by specific DataPipes' `__getstate__` and `__setstate__` ([for example](https://github.com/pytorch/pytorch/blob/gh/nivekt/45/head/torch/utils/data/datapipes/iter/combining.py#L351-L380)). IIUC you are suggesting that we should handle that here within this base class as well?

My view is that the users should 1) write their own` __getstate__` and `__setstate__` in their custom `IterDataPipe`  or 2) use `dill.dumps`.

ejguan replied:
The demux example has a problem, we are doing hack way for serialization. We would lose all the information inside the buffer as we change it to empty deque during serialization.

If we are focusing on the build-in DataPipe we provided, we will try our best to make them pickable. But, for the objects that user specified within these build-in DataPipe. as we have zero control, I think we would better to serialize by pickle first. And, if it doesn't work, we could fall back to `dill` when available or `DL2` specifies the serialization method.

NivekT replied:
Let me try something else.

NivekT replied:
For the `demux` example, I agree that the buffer will lose information. That is what I plan to address with snapshotting - the buffer's state should be restored when you restore/load a snapshot.

The various ways that I have tried to use `pickle.dumps` or `dill.dumps` within `__getstate__` are causing infinite recursion. I just tried a few different ways today but no luck. I am open to suggestion. I think it will be better for `DL2` to specify the serialization method rather than doing it within `__getstate__`.

ejguan replied:
Based on the use case of serialization, I don't think we need to implement `__getstate__` or `__setstate__`.
1. traverse with `only_datapipe=True`: no need to worry about the object 
2. traverse with `only_datapipe=False`: We could always use `dill` if available in this case 
3. multiprocessing or serialization. We could let DL2 to switch serialization method to `pickle.dumps` or `dill.dumps` on DataPipe rather than calling them within `__getstate__`.

NivekT replied:
cc: @VitalyFedyunin

NivekT commented on test/test_datapipe.py:
```diff
@@ -2015,19 +2015,23 @@ class _CustomIterDataPipe(IterDataPipe):
+        assert list(self._CustomIterDataPipe(_fake_fn)) ==\
+               list(pickle.loads(pickle.dumps(self._CustomIterDataPipe(_fake_fn))))
 
-    # TODO: Ensure this works with `dill` installed
-    # @skipIfNoDill
-    # def test_circular_serialization_with_dill(self):
-    #     assert list(self._CustomIterDataPipe()) == list(dill.loads(dill.dumps(self._CustomIterDataPipe())))
+    @skipIfNoDill
+    def test_circular_serialization_with_dill(self):
```
I am open to suggestion here to make these test cases more extensive.

ejguan replied:
Except lambda, can you add a local function here for testing?
```
def test_circular_serialization_with_dill(self):
    def _local_fn(x):
         return x + 1
    list(self._CustomIterDataPipe(_local_fn))

NivekT commented:
@ejguan I won't be working on this until next Wednesday. Feel free to push any changes as you see fit.

pmeier commented on test/test_datapipe.py:
```diff
@@ -2015,19 +2015,23 @@ class _CustomIterDataPipe(IterDataPipe):
+    @skipIfNoDill
+    def test_circular_serialization_with_dill(self):
+        assert list(self._CustomIterDataPipe(lambda x: x + 1)) == \
+               list(pickle.loads(pickle.dumps(self._CustomIterDataPipe(lambda x: x + 1))))
```
I'm a little confused. IIUC, pickling a datapipe only works with `dill`. How is this test passing?

NivekT replied:
You're right. I just double-checked and it fails, so I am removing it.

pmeier commented on test/test_datapipe.py:
```diff
@@ -2015,19 +2015,23 @@ class _CustomIterDataPipe(IterDataPipe):
         def noop(self, x):
             return x + 1
```
Nit: maybe not call it "noop" if it does something.

NivekT replied:
This has been fixed

pmeier commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -93,11 +94,10 @@ def class_function(cls, enable_df_api_tracing, source_dp, *args, **kwargs):
     def __getstate__(self):
         if IterDataPipe.getstate_hook is not None:
             return IterDataPipe.getstate_hook(self)
-        # TODO: Fix `dill` circular dependency - https://github.com/pytorch/data/issues/237
```
I don't think this can be removed. `torchvision` is going to use methods bound to `IterDataPipe`'s. Thus, if someone has `dill` installed, it still won't work, right?

NivekT replied:
It should work if this condition `if isfunction(v) and not ismethod(v):` works as I intend, because the addition of `not ismethod(v)` will prevent methods bound to `IterDataPipe` from going through the custom serialization with `dill`, which I believe will be fine as long as the methods are pickable.

NivekT committed:
Update on "[DataPipe] Restrict special serialization for non-method functions"


This PR further restricts the application of our custom serialization function in `DataPipe`. It will only be applied if:
1. `dill` is installed
2. The object in question is a non-method function

If there isn't any edge case, I believe this should address the circular dependency issue raised in https://github.com/pytorch/data/issues/237. Please let me know if you can think of anything that can potentially break this.



[ghstack-poisoned]

NivekT reviewed (commented):
I just rebased, made some changes, and added more testing. Let me know if there is a case where this may not work

NivekT commented on torch/utils/data/_utils/serialization.py:
```diff
@@ -26,11 +26,11 @@ def serialize_fn(fn):
     """
     try:
-        return pickle.dumps(fn), SerializationType("pickle")
+        return fn, SerializationType("pickle")
```
I am making this change based on @ejguan 's comment elsewhere

NivekT replied:
Fixed, I am changing it to check if `fn` is a `lambda` rather than trying to pickle `fn` with `pickle` first.

NivekT committed:
Update on "[DataPipe] Restrict special serialization for non-method functions"


This PR further restricts the application of our custom serialization function in `DataPipe`. It will only be applied if:
1. `dill` is installed
2. The object in question is a non-method function

If there isn't any edge case, I believe this should address the circular dependency issue raised in https://github.com/pytorch/data/issues/237. Please let me know if you can think of anything that can potentially break this.



[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Restrict special serialization for non-method functions"


This PR further restricts the application of our custom serialization function in `DataPipe`. It will only be applied if:
1. `dill` is installed
2. The object in question is a non-method function

If there isn't any edge case, I believe this should address the circular dependency issue raised in https://github.com/pytorch/data/issues/237. Please let me know if you can think of anything that can potentially break this.



[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Restrict special serialization for non-method functions"


This PR further restricts the application of our custom serialization function in `DataPipe`. It will only be applied if:
1. `dill` is installed
2. The object in question is a non-method lambda function

If there isn't any edge case, I believe this should address the circular dependency issue raised in https://github.com/pytorch/data/issues/237. Please let me know if you can think of anything that can potentially break this.



[ghstack-poisoned]

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2073,24 +2073,57 @@ def add_one(self, x):
+        def __init__(self, fn, source_dp=None):
+            self.fn = fn
+            self.source_dp = source_dp if source_dp else dp.iter.IterableWrapper([1, 2, 4])
+            self._dp = self.source_dp.map(self.add_one).demux(2, self.classify)[0]
```
Unrelated to pickle, this is terrible usage, as you should not pass bound methods here. The proper solution would be to have `add_one` declared as `@staticmethod`

NivekT replied:
I agree; I came up with this example to test of serialization would still work under extreme corner cases.

NivekT replied:
I will remove the bound method here

VitalyFedyunin commented:
So, the conclusion here is that nested `pickle.dumps` and `dill.dumps` as well as `dump` and similar, in any combination cause 
`RecursionError: maximum recursion depth exceeded while calling a Python object`

Minimal reproducible example here:
```python
import dill
import pickle

class B:
    def __init__(self, function):
        self.function = function

    def __getstate__(self):
        print('serializing ', self)
        r = pickle.dumps(self.function)
        print(r)
        return r

class A:
    def f_a(self, a):
        return a + 1

    def __init__(self, x):
        self.z = B(self.f_a)    
        self.val = x

a = A(3)
d = pickle.dumps(a)
```

Fighting recursion here is possible, but will require every `__getstate__` to be aware of it.

__Instead, we are choosing to prohibit the usage of `dumps` inside of the DataPipes code__ (for ourselves, for others, it is at your own risk). 

It is still ok to use `pickle` or `dill` in graph traverse and/or pipes serialization as soon as condition above is satisfied.

NivekT committed:
Update on "[DataPipe] Restrict special serialization for non-method functions"


This PR further restricts the application of our custom serialization function in `DataPipe`. It will only be applied if:
1. `dill` is installed
2. The object in question is a non-method lambda function

If there isn't any edge case, I believe this should address the circular dependency issue raised in https://github.com/pytorch/data/issues/237. Please let me know if you can think of anything that can potentially break this.



[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Restrict special serialization for non-method functions"


This PR further restricts the application of our custom serialization function in `DataPipe`. It will only be applied if:
1. `dill` is installed
2. The object in question is a non-method lambda function

If there isn't any edge case, I believe this should address the circular dependency issue raised in https://github.com/pytorch/data/issues/237. Please let me know if you can think of anything that can potentially break this.



[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111757).

NivekT committed:
Update on "[DataPipe] Revamp serialization logic of DataPipes"


This PR removes the custom serialization logic (i.e. try with `pickle` then with `dill`) inside `__getstate__` and `__setstate__` of DataPipe classes because it causes circular reference and exceptions. Instead, the attempt to use `dill` instead of `pickle` is moved to the `traversal` function for DataPipe graphs and will be included within DataLoader2 as well.

This should fix https://github.com/pytorch/data/issues/237. 

Differential Revision: [D36111757](https://our.internmc.facebook.com/intern/diff/D36111757)

[ghstack-poisoned]

NivekT commented:
I have updated this PR with the changes that we discussed offline. The only CI failure is preexisting and is due to ROCm. I have imported the PR to kick off internal tests as well.

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111757).

ejguan commented on torch/utils/data/graph.py:
```diff
@@ -19,6 +20,11 @@ def _stub_unpickler():
     f = io.BytesIO()
     p = pickle.Pickler(f)  # Not going to work for lambdas, but dill infinite loops on typing and can't be used as is
+    if DILL_AVAILABLE:
+        from dill import Pickler
+        d = Pickler(f)
```
When `dill` is imported, it will inject logic to pickler dispatch table. You need to revert it after finish traverse by doing `dill.extend(False)`

NivekT replied:
Done

NivekT committed:
Update on "[DataPipe] Revamp serialization logic of DataPipes"


This PR removes the custom serialization logic (i.e. try with `pickle` then with `dill`) inside `__getstate__` and `__setstate__` of DataPipe classes because it causes circular reference and exceptions. Instead, the attempt to use `dill` instead of `pickle` is moved to the `traversal` function for DataPipe graphs and will be included within DataLoader2 as well.

As a result, some tests need to be changed (to use `dill` explicitly when necessary) and `serialization.py` has mostly been cleaned up as the functions are no longer needed.

This should fix https://github.com/pytorch/data/issues/237. 

Differential Revision: [D36111757](https://our.internmc.facebook.com/intern/diff/D36111757)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111757).

ejguan commented on torch/utils/data/graph.py:
```diff
@@ -46,14 +52,23 @@ def reduce_hook(obj):
+        except (pickle.PickleError, AttributeError, TypeError):
+            if DILL_AVAILABLE:
+                d.dump(scan_obj)
+            else:
+                pass
```
Wait, I don't think we should pass here. If the `DataPipe` can not be `dump`ed, we can't save it to the graph. So, a proper Error should raise to indicate that.

NivekT replied:
You are right, I am changing `pass` to `raise`

NivekT replied:
I suspect it may be caught by the next line (62-63), unless we want to change that right now

ejguan replied:
I think we should remove 62-63 as well. wdyt

NivekT replied:
Yea, I think it is better for the users/us to know if something is not pickable rather than silently failing and not capturing it within the graph.

ejguan replied:
IIRC, this traverse function is only used by `DataLoader2`? If `DataPipe` is provided to `DataLoader`, it won't check if the `DataPipe` graph is traversable or not. A proper error would raise when multiprocessing involved.

NivekT replied:
`traverse` is used by `apply_shuffle_settings`, which is used here within DataLoader:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L245

ejguan replied:
Thank you for clarification. I still think it's better to remove the logic to suppressing pickling Error as you pointed out.

> I think it is better for the users/us to know if something is not pickable rather than silently failing and not capturing it within the graph.

NivekT replied:
Agree, I am updating this now.

NivekT committed:
Update on "[DataPipe] Revamp serialization logic of DataPipes"


This PR removes the custom serialization logic (i.e. try with `pickle` then with `dill`) inside `__getstate__` and `__setstate__` of DataPipe classes because it causes circular reference and exceptions. Instead, the attempt to use `dill` instead of `pickle` is moved to the `traversal` function for DataPipe graphs and will be included within DataLoader2 as well.

As a result, some tests need to be changed (to use `dill` explicitly when necessary) and `serialization.py` has mostly been cleaned up as the functions are no longer needed.

This should fix https://github.com/pytorch/data/issues/237. 

Differential Revision: [D36111757](https://our.internmc.facebook.com/intern/diff/D36111757)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111757).

ejguan commented on test/test_datapipe.py:
```diff
@@ -2067,43 +2083,127 @@ def test_traverse_mixdatapipe(self):
 
     class CustomIterDataPipe(IterDataPipe):
-        def add_one(self, x):
+
+        @staticmethod
+        def add_one(x):
```
We still need a function attached to this instance because the function might needs some info from `self` like:
```py
class CustomIterDataPipe(IterDataPipe):
    def __init__(self, ...):
        self.v = 1
    def add_v(self, x):
        return x + self.v
```

ejguan replied:
This is the corner case we want to cover.

NivekT replied:
Do you intend to pass `add_v` to say `self.dp.map(add_v)`? Or just what you typed out?

May be related:
https://github.com/pytorch/pytorch/pull/74984#discussion_r863292137

NivekT replied:
In any case, I can add that to the test

NivekT committed:
Update on "[DataPipe] Revamp serialization logic of DataPipes"


This PR removes the custom serialization logic (i.e. try with `pickle` then with `dill`) inside `__getstate__` and `__setstate__` of DataPipe classes because it causes circular reference and exceptions. Instead, the attempt to use `dill` instead of `pickle` is moved to the `traversal` function for DataPipe graphs and will be included within DataLoader2 as well.

As a result, some tests need to be changed (to use `dill` explicitly when necessary) and `serialization.py` has mostly been cleaned up as the functions are no longer needed.

This should fix https://github.com/pytorch/data/issues/237. 

Differential Revision: [D36111757](https://our.internmc.facebook.com/intern/diff/D36111757)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111757).

ejguan commented on test/test_datapipe.py:
```diff
@@ -2065,45 +2084,156 @@ def test_traverse_mixdatapipe(self):
+        self.assertTrue(str(dp2) in str(res3))  # Quick check to ensure the result isn't blank
+        self.assertTrue(str(dp2) in str(res4))
+
+    class LambdaIterDataPipe(IterDataPipe):
```
Why don't we have a single `LambdaIterDataPipe` for testing? IIUC, `LambdaIterDdataPipe` is a superset of `CustomIterDataPipe`.

NivekT replied:
I refactored the class. Thanks for pointing that out!

ejguan reviewed (approved):
Thank you so much. LGTM with one comment on the test

NivekT committed:
Update on "[DataPipe] Revamp serialization logic of DataPipes"


This PR removes the custom serialization logic (i.e. try with `pickle` then with `dill`) inside `__getstate__` and `__setstate__` of DataPipe classes because it causes circular reference and exceptions. Instead, the attempt to use `dill` instead of `pickle` is moved to the `traversal` function for DataPipe graphs and will be included within DataLoader2 as well.

As a result, some tests need to be changed (to use `dill` explicitly when necessary) and `serialization.py` has mostly been cleaned up as the functions are no longer needed.

This should fix https://github.com/pytorch/data/issues/237. 

Differential Revision: [D36111757](https://our.internmc.facebook.com/intern/diff/D36111757)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111757).
